driver-core/src/main/com/mongodb/internal/connection/OidcAuthenticator.java

```diff
@@ -0,0 +1,582 @@
+            } catch (MongoCommandException e) {
+                if (triggersRetry(e)) {
+                    authLock(connection, connectionDescription);
+                }
+            }
+        } else {
+            authLock(connection, connectionDescription);
+        }
+    }
+
+    private static boolean triggersRetry(@Nullable final Throwable t) {
+        if (t instanceof MongoSecurityException) {
+            MongoSecurityException e = (MongoSecurityException) t;
+            Throwable cause = e.getCause();
+            if (cause instanceof MongoCommandException) {
+                MongoCommandException commandCause = (MongoCommandException) cause;
+                return commandCause.getErrorCode() == 18;
+            }
+        }
+        return false;
+    }
+
+
+    private void authenticateUsing(
+            final InternalConnection connection,
+            final ConnectionDescription connectionDescription,
+            final Function<byte[], byte[]> evaluateChallengeFunction) {
+        this.evaluateChallengeFunction = evaluateChallengeFunction;
+        super.authenticate(connection, connectionDescription);
+    }
+
+    private void authLock(final InternalConnection connection, final ConnectionDescription connectionDescription) {
+        MongoCredentialWithCache mongoCredentialWithCache = getMongoCredentialWithCache();
+        mongoCredentialWithCache.withOidcLock(() -> {
+            while (true) {
+                try {
+                    authenticateUsing(connection, connectionDescription, (challenge) -> evaluate(challenge));
+                    break;
+                } catch (MongoCommandException | MongoSecurityException e) {
+                    OidcCacheEntry cacheEntry = mongoCredentialWithCache.getOidcCacheEntry();
+                    if (triggersRetry(e)) {
+                        prepareRetry(e, cacheEntry);
+                    } else {
+                        throw e;
+                    }
+                }
+            }
+            return null;
+        });
+    }
+
+    private byte[] evaluate(final byte[] challenge) {
+        OidcRequestCallback requestCallback = getRequestCallback();
+        if (requestCallback == null) {
+            return sendAwsFromFile();
+        }
+
+        MongoCredentialWithCache mongoCredentialWithCache = getMongoCredentialWithCache();
+        OidcCacheEntry cacheEntry = mongoCredentialWithCache.getOidcCacheEntry();
+        String cachedAccessToken = getValidCachedAccessToken();
+        String invalidConnectionAccessToken = connectionLastAccessToken;
+        String cachedRefreshToken = cacheEntry.refreshToken;
+        IdpServerInfo cachedIdpServerInfo = cacheEntry.idpServerInfo;
+
+        if (cachedAccessToken != null) {
+            boolean cachedTokenIsInvalid = cachedAccessToken.equals(invalidConnectionAccessToken);
+            if (cachedTokenIsInvalid) {
+                mongoCredentialWithCache.setOidcCacheEntry(cacheEntry.clearAccessToken());
```

Comment:
As far as I can see, this "check-then-setOidcCacheEntry" is allowed to happen concurrently with other invocations of `setOidcCacheEntry` as follows:

* `InternalStreamConnectionInitializer.createHelloCommand`
  * `OidcAuthenticator.createSpeculativeAuthenticateCommand`
    * `OidcAuthenticator.evaluate`

I.e., this "check-then-setOidcCacheEntry" is not atomic with respect to other actions involving `setOidcCacheEntry`. I don't think this is correct.

Reply:
In the latest code, setOidcCacheEntry is only called by code that is ultimately wrapped in withOidcLock. Possibly this was broken before I updated the code with the latest spec change.

Reply:
> In the latest code

Do you mean commit d494ea1d70f32de4534f28c735f557a5569cd0f5 (the latest commit in the PR) or your local code? The code that is in the PR contains "check-then-setOidcCacheEntry" in `OidcAuthenticator.evaluate`, which is called without holding the lock in the call chain I specified above. Maybe I am missing the code that acquires the lock in this chain, in that case could you please point me to that code?

Reply:
I meant local, apologies. I was unable to find the string `check-then-setOidcCacheEntry` so I believe this is fixed. `evaluate` is always called under lock.